// nuclide-commons/lib/clientInfo.js
Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.isRunningInTest = isRunningInTest;
exports.isRunningInClient = isRunningInClient;
exports.getAtomVersion = getAtomVersion;
exports.isRunningInNuclide = isRunningInNuclide;
exports.getNuclideBuildNumber = getNuclideBuildNumber;
exports.getInstallerPackageBuildNumber = getInstallerPackageBuildNumber;
exports.setInstallerPackageBuildNumber = setInstallerPackageBuildNumber;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _singleton = require('./singleton');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

'use babel';

var SMALLEST_NUCLIDE_BUILD_NUMBER = 5394875;
var INSTALLER_BUILD_NUMBER_KEY = '_nuclide_installer_build_number_key';

function isRunningInTest() {
  return process.env.NODE_ENV === 'test';
}

function isRunningInClient() {
  return global.atom !== undefined;
}

function getAtomVersion() {
  if (!isRunningInClient()) {
    throw Error('Not running in Atom/Nuclide.');
  }
  return global.atom.getVersion();
}

/**
 * Determine whether the package is running in Atom.app or Nuclide.app.
 *
 * While building Nuclide release, we set the internal build number as part of version number.
 * So the version looks like 1.0.$buildNumber. Since the internal build number is really big,
 * whereas the counterpart in Atom' version string is realtively small or not a valid number,
 * it is a good way to identify if the running editer is Nuclide or Atom.
 */

function isRunningInNuclide() {
  if (!isRunningInClient()) {
    return false;
  }

  var version = getAtomVersion();
  var buildNumber = version.split('.')[2];
  // If the PATCH version (the third part of version string splitted by dot) is a number and larger
  // than SMALLEST_NUCLIDE_BUILD_NUMBER, then it's a build number.
  if (/^\d+$/.test(buildNumber) && parseInt(buildNumber, 10) >= SMALLEST_NUCLIDE_BUILD_NUMBER) {
    return true;
  }
  return false;
}

var atomConfig = isRunningInClient() ? require(_path2['default'].join(atom.getLoadSettings().resourcePath, 'package.json')) : {};

function getNuclideBuildNumber() {
  if (!isRunningInClient()) {
    throw Error('Not running in Atom/Nuclide.');
  }
  return atomConfig.buildNumber || 0;
}

function getInstallerPackageBuildNumber() {
  return (0, _singleton.get)(INSTALLER_BUILD_NUMBER_KEY, function () {
    return 0;
  });
}

function setInstallerPackageBuildNumber(buildNumber) {
  (0, _singleton.reset)(INSTALLER_BUILD_NUMBER_KEY, function () {
    return buildNumber;
  });
}

// TODO(chenshen) implement isDevelopment.

var __test__ = {
  SMALLEST_NUCLIDE_BUILD_NUMBER: SMALLEST_NUCLIDE_BUILD_NUMBER
};
exports.__test__ = __test__;
