// nuclide-commons/lib/runtimeInfo.js
Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.getRuntimeInformation = getRuntimeInformation;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _object = require('./object');

var _clientInfo = require('./clientInfo');

var _systemInfo = require('./systemInfo');

var _environment = require('./environment');

var _environment2 = _interopRequireDefault(_environment);

var _session = require('./session');

var _session2 = _interopRequireDefault(_session);

'use babel';

var cachedInformation = null;

function getCacheableRuntimeInformation() {
  if (cachedInformation !== null) {
    return cachedInformation;
  }

  cachedInformation = {
    sessionId: '',
    user: _environment2['default'].USER,
    osType: (0, _systemInfo.getOsType)(),
    timestamp: 0,
    isClient: (0, _clientInfo.isRunningInClient)(),
    atomVersion: (0, _clientInfo.isRunningInClient)() ? (0, _clientInfo.getAtomVersion)() : '',
    isRunningInNuclide: (0, _clientInfo.isRunningInNuclide)(),
    nuclideVersion: (0, _clientInfo.isRunningInNuclide)() ? (0, _clientInfo.getNuclideBuildNumber)() : 0,
    installerPackageVersion: 0,
    uptime: 0,
    // TODO (chenshen) fill following information.
    serverVersion: 0,
    isDevelopment: false
  };

  return cachedInformation;
}

function getRuntimeInformation() {
  var runtimeInformation = (0, _object.assign)({}, getCacheableRuntimeInformation());
  runtimeInformation.sessionId = _session2['default'].id;
  runtimeInformation.timestamp = Date.now();
  runtimeInformation.installerPackageVersion = (0, _clientInfo.getInstallerPackageBuildNumber)();
  runtimeInformation.uptime = Math.floor(process.uptime() * 1000);
  return runtimeInformation;
}
